File: 2023/01-trebuchet/part1.lua
local fp = assert(io.open(arg[1] or "./input.txt", "r"))
local count = 1
local sum = 0

---@param fmt string
---@param ... string|number
local function printf(fmt, ...)
    io.stdout:write(fmt:format(...))
end

---@param line string
local function find_numbers(line)
    -- start of string capture nearest digit, non-greedy skip non-digits before us
    local _, _, left = line:find("^[%D]-(%d)")
    -- end of string capture nearest digit, non-greedy skip non-digits after us
    local _, _, right = line:find("(%d)[%D]-$")
    return left, right
end

for line in fp:lines("*l") do
    if (line ~= "") then 
        local left, right = find_numbers(line)
        printf("%i : '%s', Digits: %i, %i\n", count, line, left, right)
        count = count + 1
        -- We want to treat 1 and 2 as 12, 7 and 7 as 77, etc.
        sum = sum + (left * 10) + right;
    end
end
printf("Sum: %i\n", sum)

fp:close()

File: 2023/01-trebuchet/c/part1.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFFERSIZE 256 // totally safe!

#define eprintf(msg) fprintf(stderr, __FILE__ ":%i: " msg "\n", __LINE__)

// Totally safe!
int find_number(const char *buffer, int increment) {
    char c = -1;
    while (!isdigit(c)) {
        c = *buffer;
        buffer += increment;
    }
    // If didn't find a number, default to 0
    return (c > 0) ? c - '0' : 0;
}

// Gets the first and last digit of `buffer`
int match_numbers(const char *buffer, const int length) {
    // Avoid working with empty strings
    if (*buffer == '\0') {
        return 0;
    }
    // From start of string, try to find a number to the right
    int x = find_number(buffer, 1);
    // Get pointer to end of string, try find a number to the left 
    int y = find_number(buffer + length, -1);
    return (x * 10) + y;
}

int main(int argc, char *argv[]) {
    // totally safe!
    const char *name = (argc == 2) ? argv[1] : "../part1.txt";
    FILE *file = fopen(name, "r");
    if (!file) {
        eprintf("Failed to open file!");
        return 1;
    }
    // Pray we don't overflow this :)
    char buffer[BUFFERSIZE] = {0};
    int count = 1;
    int sum = 0;
    // Return NULL on EOF
    while (fgets(buffer, BUFFERSIZE, file) != NULL) {
        // Get index of first occurence of some newline char
        buffer[strcspn(buffer, "\r\n")] = '\0';
        int matched = match_numbers(buffer, strlen(buffer));
        printf("%i: %s\t%i\n", count++, buffer, matched);
        sum += matched;
    }
    printf("Calibration Value: %i\n", sum);
    fclose(file);
    return 0;
}
